---
title: "HMP16SData"
author:
- name: Lucas Schiffer
  affiliation:
  - &1 Graduate School of Public Health and Health Policy, City University of
    New York, New York, NY
  - &2 Institute for Implementation Science in Population Health, City
    University of New York, New York, NY
- name: Rimsha Azhar
  affiliation:
  - *1
  - *2
- name: Marcel Ramos
  affiliation:
  - *1
  - *2
  - &3 Roswell Park Cancer Institute, University of Buffalo, Buffalo, NY
- name: Ludwig Geistlinger
  affiliation:
  - *1
  - *2
- name: Levi Waldron
  affiliation:
  - *1
  - *2
date: '`r format(Sys.Date(), "%B %e, %Y")`'
abstract: >
    HMP16SData is a Bioconductor ExperimentData package of the Human Microbiome
    Project (HMP) 16S rRNA sequencing data for variable regions 1–3 and 3–5. Raw
    data files are provided in the package as downloaded from the HMP Data
    Analysis and Coordination Center. Processed data is provided as
    SummarizedExperiment class objects via ExperimentHub.
package: HMP16SData
output:
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{HMP16SData}
    %\VignetteEncoding{UTF-8}
    %\VignetteEngine{knitr::rmarkdown}
editor_options:
    chunk_output_type: console
---

# Prerequisites

The following `r BiocStyle::CRANpkg("knitr")` options will be used in this
vignette to provide the most useful and concise output.

```{r}
knitr::opts_chunk$set(message = FALSE)
```

The following packages will be used in this vignette to provide demonstrative
examples of what a user might do with `r BiocStyle::Biocpkg("HMP16SData")`.

```{r}
library(HMP16SData)
library(phyloseq)
library(magrittr)
library(dplyr)
library(ggplot2)
library(curatedMetagenomicData)
library(gridExtra)
library(cowplot)
```

Pipe operators from the `r BiocStyle::CRANpkg("magrittr")` package are used in
this vignette to provide the most elegant and concise syntax. See the
`r BiocStyle::CRANpkg("magrittr")` vignette if the syntax is unclear.

# Introduction

`r BiocStyle::Biocpkg("HMP16SData")` is a Bioconductor ExperimentData package of
the Human Microbiome Project (HMP) 16S rRNA sequencing data for variable regions
1–3 and 3–5. Raw data files are provided in the package as downloaded from the
[HMP Data Analysis and Coordination Center](https://tinyurl.com/y7ev836z).
Processed data is provided as `SummarizedExperiment` class objects via
`r BiocStyle::Biocpkg("ExperimentHub")`.

`r BiocStyle::Biocpkg("HMP16SData")` can be installed using
`r BiocStyle::Biocpkg("BiocInstaller")` as follows.

```{r, eval=FALSE}
BiocInstaller::biocLite("HMP16SData")
```

Once installed, `r BiocStyle::Biocpkg("HMP16SData")` provides two functions to
access data – one for variable region 1–3 and another for variable region 3–5.
When called, as follows, the functions will download data from an
`r BiocStyle::Biocpkg("ExperimentHub")` Amazon S3 (Simple Storage Service)
bucket over `https` or load data from a local cache.

```{r}
V13()
V35()
```

The two data sets are represented as `SummarizedExperiment` objects, a standard
Bioconductor class that is amenable to subsetting and analysis. To maintain
brevity, details of the `SummarizedExperiment` class are not outlined here but
the `r BiocStyle::Biocpkg("SummarizedExperiment")` package provides an excellent
vignette.

# Features

## Frequency Table Generation

Sometimes it is desirable to provide a quick summary of key demographic
variables and to make the process easier `r BiocStyle::Biocpkg("HMP16SData")`
provides a function, `table_one`, to do so. It returns a `data.frame` or a
`list` of `data.frame` objects that has been transformed to make a publication
ready table.

```{r}
V13() %>%
    table_one() %>%
    head()
```

If a `list` is passed to `table_one`, its elements must be named so that the
named elements can be used by the `kable_one` function. The `kable_one` function
will produce an `HTML` table for vignettes such as the one shown below.

```{r}
list(V13 = V13(), V35 = V35()) %>%
    table_one() %>%
    kable_one()
```

## Straightforward Subsetting

The `SummarizedExperiment` container provides for straightforward subsetting by
data or metadata variables using either the `subset` function or `[` methods –
see the `r BiocStyle::Biocpkg("SummarizedExperiment")` vignette for additional
details. Shown below, the variable region 3–5 data set is subset to include only
stool samples.

```{r}
V35_stool <-
    V35() %>%
    subset(select = HMP_BODY_SUBSITE == "Stool")

V35_stool
```

## Protected Metadata Attachment

The HMP study has extensive protected metadata available through the National
Center for Biotechnology Information (NCBI) database of Genotypes and Phenotypes
(dbGaP) that can readily be attached to the `SummarizedExperiment` objects. Only
a dbGaP repository key is required to download, process, and attach the
protected metadata. See <https://dbgap.ncbi.nlm.nih.gov/aa> for information
about the protected metadata available and obtaining a key. The `attach_dbGap()`
function takes a `r BiocStyle::Biocpkg("HMP16SData")` `SummarizedExperiment`
object as its first argument and the path to the user's dbGaP repository key as
its second argument, and it returns another `SummarizedExperiment` also
containing the protected metadata.

```{r, eval=FALSE}
V35_stool_protected <-
    V35_stool %>%
    attach_dbGaP("~/prj_12146.ngc")
```

The returned `SummarizedExperiment` object has protected metadata as additional
columns in the `colData` – all metadata can be returned as follows.

```{r, eval=FALSE}
colData(V35_stool_protected)
```

Descriptions of the variables attached using the `attach_dbGaP` function, as
translated from dbGaP `XLM` data dictionary files, are available in the
`dictionary` object.

```{r}
head(dictionary)
```

## Phyloseq Class Coercion

Data from the HMP study is likely useful as a comparison or control group, a
*"standard candle"* of sorts, in the context of other human microbiome studies.
As such, `r BiocStyle::Biocpkg("HMP16SData")` provides a method, `as_phyloseq`,
to coerce `SummarizedExperiment` class objects to `phyloseq` class objects. The
`r BiocStyle::Biocpkg("phyloseq")` package provides an extensive suite of
methods to analyze microbiome data and many potential comparisons could be made.

```{r}
V35_stool_phyloseq <-
    V35_stool %>%
    as_phyloseq()
```

When plotting `phyloseq` objects it is necessary to remove taxa that sum to zero
(i.e. have no relative abundance) to avoid warning messages.

```{r}
V35_stool_phyloseq %<>%
    taxa_sums() %>%
    is_greater_than(0) %>%
    prune_taxa(V35_stool_phyloseq)
```

With zero-abundance taxa pruned, exploratory plots can easily be made. For
example, a notched box plot of three alpha diversity measures can be made in a
few lines.

```{r, fig.height=5, fig.width=8}
richness_measures <- c("Observed", "Shannon", "Simpson")

V35_stool_phyloseq %>%
    plot_richness(x = "SEX", color = "SEX", measures = richness_measures) +
    stat_boxplot(geom ="errorbar") +
    geom_boxplot(notch = TRUE) +
    theme_bw() +
    theme(axis.title.x = element_blank(), legend.position = "none")
```

# External Data Comparison

In addition to 16S rRNA sequencing, the HMP Study conducted whole metagenome
shotgun (WMS) sequencing but the samples are present in another package,
`r BiocStyle::Biocpkg("curatedMetagenomicData")`. This type of external data
comparison is likely a common use case and a phylum-level relative abundance
comparison of the 16S and WMS samples is made here to illustrate how an external
data comparison might be made.

The `V35_stool_phyloseq` object constructed above contains the 16S variable
region 3–5 data subset to include only stool samples, but the WMS stool samples
are obtained and coerced to a `phyloseq` object as follows.

```{r}
WMS_stool_phyloseq <-
    HMP_2012.metaphlan_bugs_list.stool() %>%
    ExpressionSet2phyloseq()
```

The `r BiocStyle::Biocpkg("phyloseq")` package can then be used to summarize
abundance counts at the phylum level using the `tax_glom` function. Thereafter,
given the complexity of operations needed to make a phylum-level relative
abundance plot, the `phyloseq` class objects are *"melted"* into `data.frame`
class objects using the `psmelt` function.

```{r}
V35_stool_melted <-
    V35_stool_phyloseq %>%
    tax_glom(taxrank = "PHYLUM") %>%
    psmelt()

WMS_stool_melted <-
    WMS_stool_phyloseq %>%
    tax_glom(taxrank = "Phylum") %>%
    psmelt()
```

There is a column `SRS_SAMPLE_ID` present in the 16S variable region 3–5 data
that is explicitly for mapping to the WMS samples and the matching identifiers
are in the `NCBI_accession` column of the WMS sample data. The intersection of
the two vectors represents the matching samples that are present in both data
sets. This intersection, shown below as `SRS_SAMPLE_IDS`, can then be used to
filter both the 16S and WMS samples to include only the samples in common.

Along with the `filter` step shown below, standardization of sample identifiers
to the `SRS` numbers and conversion of taxonomy counts to relative abundance is
done. The conversion to relative abundance is necessary for comparability across
the 16S and WMS samples.

In either case, data is first grouped by samples and the percent composition of
each phylum relative to the others in the sample is calculated by taking the
count abundance of the phylum divided by the sum of all abundance counts in the
sample. Samples are then sorted by phylum and descending relative abundance
before being grouped by phylum – the grouping by phylum allows for the
assignment a phylum rank that is then used to sort the phyla by descending
relative abundance. Finally, only the sample, phylum, and relative abundance
columns are kept once the order has been established and all groupings can be
removed.

```{r}
SRS_SAMPLE_IDS <-
    intersect(V35_stool_melted$SRS_SAMPLE_ID, WMS_stool_melted$NCBI_accession)

V35_stool_melted %<>%
    filter(SRS_SAMPLE_ID %in% SRS_SAMPLE_IDS) %>%
    rename(Phylum = PHYLUM) %>%
    mutate(Sample = SRS_SAMPLE_ID) %>%
    group_by(Sample) %>%
    mutate(`Relative Abundance` = Abundance / sum(Abundance)) %>%
    arrange(Phylum, -`Relative Abundance`) %>%
    group_by(Phylum) %>%
    mutate(phylum_rank = sum(`Relative Abundance`)) %>%
    arrange(-phylum_rank) %>%
    select(Sample, Phylum, `Relative Abundance`) %>%
    ungroup()

WMS_stool_melted %<>%
    filter(NCBI_accession %in% SRS_SAMPLE_IDS) %>%
    group_by(Sample) %>%
    mutate(`Relative Abundance` = Abundance / sum(Abundance)) %>%
    arrange(Phylum, -`Relative Abundance`) %>%
    group_by(Phylum) %>%
    mutate(phylum_rank = sum(`Relative Abundance`)) %>%
    arrange(-phylum_rank) %>%
    select(Sample, Phylum, `Relative Abundance`) %>%
    ungroup()
```

Provided that the phyla are ordered by relative abundance, the top phyla from
each data set can be obtained and intersected to give the top eight phyla
present in both data sets. The top eight phyla can then be used to filter 16S
and WMS samples to include only the desired phyla.

```{r}
V35_top_phyla <-
    V35_stool_melted %$%
    unique(Phylum) %>%
    as.character()

WMS_top_phyla <-
    WMS_stool_melted %$%
    unique(Phylum) %>%
    as.character()

top_eight_phyla <-
    intersect(V35_top_phyla, WMS_top_phyla) %>%
    extract(1:8)

V35_stool_melted %<>%
    filter(Phylum %in% top_eight_phyla)

WMS_stool_melted %<>%
    filter(Phylum %in% top_eight_phyla)
```

To achieve ordering of samples by the relative abundance of the top phylum when
plotting, a vector of the unique sample identifiers is constructed and will be
used as the `levels` of a `factor`.

```{r}
sample_order <-
    V35_stool_melted %$%
    unique(Sample)
```

A vector of unique phyla is also constructed and will be used as the `levels` of
a `factor` when plotting. If this were not done, phyla would simply be sorted
alphabetically.

```{r}
phylum_order <-
    V35_stool_melted %$%
    unique(Phylum)
```

Color blindness affects a significant portion of the population, but, using an
intelligent color pallet, figures can be designed to be friendly to those with
deuteranopia, protanopia, and tritanopia. The following colors are derived from
Wong, B. Points of view: Color blindness. *Nat. Methods* **8**, 441 (2011).

```{r}
bang_wong_colors <-
    c(
        "#CC79A7",
        "#D55E00",
        "#0072B2",
        "#F0E442",
        "#009E73",
        "#56B4E9",
        "#E69F00",
        "#000000"
    )
```

Using the `sample_order` and `phylum_order` vectors constructed above, stacked
phylum-level relative abundance bar plots sorted by decreasing relative
abundance of the top phylum and stratified by the top eight phyla can be made
for 16S and WMS samples. The two plots are made separately and serialized so
that they can be arranged in a single figure for comparison.

```{r}
V35_plot <-
    V35_stool_melted %>%
    mutate(Sample = factor(Sample, sample_order)) %>%
    mutate(Phylum = factor(Phylum, phylum_order)) %>%
    ggplot(aes(Sample, `Relative Abundance`, fill = Phylum)) +
    geom_bar(stat = "identity", position = "fill", width = 1) +
    scale_fill_manual(values = bang_wong_colors) +
    theme_minimal() +
    theme(axis.text.x = element_blank(), axis.title.x = element_blank(),
          panel.grid = element_blank(), legend.position = "none",
          legend.title = element_blank()) +
    ggtitle("Phylum-Level Relative Abundance", "16S Stool Samples")

WMS_plot <-
    WMS_stool_melted %>%
    mutate(Sample = factor(Sample, sample_order)) %>%
    mutate(Phylum = factor(Phylum, phylum_order)) %>%
    ggplot(aes(Sample, `Relative Abundance`, fill = Phylum)) +
    geom_bar(stat = "identity", position = "fill", width = 1) +
    scale_fill_manual(values = bang_wong_colors) +
    theme_minimal() +
    theme(axis.text.x = element_blank(), axis.title.x = element_blank(),
          panel.grid = element_blank(), legend.position = "none",
          legend.title = element_blank()) +
    ggtitle("", "WMS Stool Samples")
```

In the plots above, legends are removed to reduce redundancy, but a legend is
still necessary and can be serialized as its own plot below using the
`get_legend` function from the `r BiocStyle::CRANpkg("cowplot")` package.

```{r}
plot_legend <- {
        WMS_plot +
            theme(legend.position = "bottom")
    } %>%
    get_legend()
```

Finally, the `grid.arrange` function from the
`r BiocStyle::CRANpkg("gridExtra")` package is used to arrange, scale, and plot
the three plots in a single figure.

```{r, fig.height=8, fig.width=8}
grid.arrange(V35_plot, WMS_plot, plot_legend, ncol = 1, heights = c(3, 3, 1))
```

Notably, the figure illustrates the Bacteroidetes/Firmicutes gradient with some
discretion between the 16S and WMS samples.

# Limitations

Data from variable regions 1–3 and 3–5 are not comparable to everything, be
careful with assumptions and conduct analysis that is methodologically sound.

# Session Information

```{r}
sessionInfo()
```
