---
title: "HMP16SData"
date: '`r format(Sys.Date(), "%B %e, %Y")`'
author:
    - Lucas Schiffer
    - Rimsha Azhar
    - Marcel Ramos
    - Ludwig Geistlinger
    - Levi Waldron
package: HMP16SData
output:
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{HMP16SData}
    %\VignetteEncoding{UTF-8}
    %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

# Prerequisites

The following packages will be used in this vignette to provide demonstrative
examples of what a user might do with `r BiocStyle::Biocpkg("HMP16SData")`. The
`r BiocStyle::Githubpkg("schifferl/kableOne")` package is available only through
GitHub at present.

```{r, message=FALSE}
BiocInstaller::biocLite("schifferl/kableOne")

library(HMP16SData)
library(phyloseq)
library(kableOne)
```

# Introduction

`r BiocStyle::Biocpkg("HMP16SData")` is a Bioconductor ExperimentHub package of
the Human Microbiome Project (HMP) 16S ribosomal RNA (rRNA) sequencing data for
variable regions 1-3 and 3-5. The raw data files are provided in the package as
downloaded from the HMP Data Analysis and Coordination Center (DACC) website. 
Processed data is provided as `SummarizedExperiment` objects via
`ExperimentHub`.

`r BiocStyle::Biocpkg("HMP16SData")` can be installed using 
`r BiocStyle::Biocpkg("BiocInstaller")` as follows:

```{r, eval=FALSE}
BiocInstaller::biocLite("HMP16SData")
```

Once installed, `r BiocStyle::Biocpkg("HMP16SData")` provides two functions to 
access data - one for variable region 1-3 and another for variable region 3-5. 
When called, as follows, the functions will download data from an 
`r BiocStyle::Biocpkg("ExperimentHub")` Amazon S3 (Simple Storage Service) 
bucket over `https`.

```{r, message=FALSE}
V13()
V35()
```

The two datasets are represented as `SummarizedExperiment` objects, a standard 
Bioconductor class that is amenable to subsetting and analysis. To maintain 
breveity, details of the `SummarizedExperiment` class are not outlined here but 
the `r BiocStyle::Biocpkg("SummarizedExperiment")` package provides an excellent
vignette.

# Features

## Straightforward Subsetting

The `SummarizedExperiment` container provides for straightforward subsetting by
data or metadata variables using either the `subset` function or `[` methods -
see the `r BiocStyle::Biocpkg("SummarizedExperiment")` vignette for additional
details. Shown below, the variable region 1-3 dataset is subset to include only
stool samples.

```{r, message=FALSE}
V13_stool <-
    V13() %>%
    subset(select = hmp_body_subsite == "Stool")

V13_stool
```

## Protected Metadata Attachment

The HMP study has extensive protected metadata available through the National
Center for Biotechnology Information's (NCBI) database of Genotypes and
Phenotypes (dbGaP) that can readily be attached to the `SummarizedExperiment`
objects. Only a dbGaP repository key is required to download, process, and
attach the protected metadata - an unevaluated example is shown below for those
with access to protected metadata. See <https://dbgap.ncbi.nlm.nih.gov/aa> for
additional details.

```{r, eval=FALSE}
V13_stool %>%
    attach_dbGaP("~/prj_12146.ngc") %>%
    as.phyloseq() %>%
    plot_richness(x = "sex", color = "sex", measures = "Shannon")
```

## Taxonomic Identifier Translation

```{r, eval=FALSE}
V13_stool %>%
    translate_identifiers()
```

## External Data Comparison

Topic to be covered in a searate vignette.

## Frequency Table Generation

Be aware that the `kable_one` function from the
`r BiocStyle::Githubpkg("schifferl/kableOne")` package is _very_ beta.

```{r, message=FALSE}
V13 <-
    V13() %>%
    table_one()

V35 <-
    V35() %>%
    table_one()

kable_one(V13, V35)
```

# Limitations

Data from variable regions 1-3 and 3-5 are not comparable to everything.
