---
title: "HMP16SData"
date: '`r format(Sys.Date(), "%B %e, %Y")`'
author:
    - Lucas Schiffer
    - Rimsha Azhar
    - Marcel Ramos
    - Ludwig Geistlinger
    - Levi Waldron
package: HMP16SData
abstract: >
    The HMP16SData package is a Bioconductor ExperimentHub package of the Human
    Microbiome Project (HMP) 16S rRNA sequencing data for variable regions 1–3
    and 3–5. The raw data files are provided in the package as downloaded from
    the HMP Data Analysis and Coordination Center website. Processed data is
    provided as SummarizedExperiment objects via ExperimentHub.
output:
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{HMP16SData}
    %\VignetteEncoding{UTF-8}
    %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

# Prerequisites

The following packages will be used in this vignette to provide demonstrative
examples of what a user might do with `r BiocStyle::Biocpkg("HMP16SData")`. The
`r BiocStyle::Githubpkg("schifferl/kableOne")` package is available only through
GitHub at present.

```{r, message=FALSE}
BiocInstaller::biocLite("schifferl/kableOne")

library(HMP16SData)
library(phyloseq)
library(magrittr)
library(ggplot2)
library(kableOne)
```

# Introduction

`r BiocStyle::Biocpkg("HMP16SData")` is a Bioconductor ExperimentHub package of
the Human Microbiome Project (HMP) 16S ribosomal RNA (rRNA) sequencing data for
variable regions 1–3 and 3–5. The raw data files are provided in the package as
downloaded from the HMP Data Analysis and Coordination Center (DACC) website. 
Processed data is provided as `SummarizedExperiment` objects via ExperimentHub.

`r BiocStyle::Biocpkg("HMP16SData")` can be installed using 
`r BiocStyle::Biocpkg("BiocInstaller")` as follows:

```{r, eval=FALSE}
BiocInstaller::biocLite("HMP16SData")
```

Once installed, `r BiocStyle::Biocpkg("HMP16SData")` provides two functions to 
access data – one for variable region 1–3 and another for variable region 3–5. 
When called, as follows, the functions will download data from an 
`r BiocStyle::Biocpkg("ExperimentHub")` Amazon S3 (Simple Storage Service) 
bucket over `https`.

```{r, message=FALSE}
V13()
V35()
```

The two datasets are represented as `SummarizedExperiment` objects, a standard 
Bioconductor class that is amenable to subsetting and analysis. To maintain 
breveity, details of the `SummarizedExperiment` class are not outlined here but 
the `r BiocStyle::Biocpkg("SummarizedExperiment")` package provides an excellent
vignette.

# Features

## Straightforward Subsetting

The `SummarizedExperiment` container provides for straightforward subsetting by
data or metadata variables using either the `subset` function or `[` methods –
see the `r BiocStyle::Biocpkg("SummarizedExperiment")` vignette for additional
details. Shown below, the variable region 1–3 dataset is subset to include only
stool samples.

```{r, message=FALSE}
V13_stool <-
    V13() %>%
    subset(select = hmp_body_subsite == "Stool")

V13_stool
```

## Protected Metadata Attachment

The HMP study has extensive protected metadata available through the National
Center for Biotechnology Information's (NCBI) database of Genotypes and
Phenotypes (dbGaP) that can readily be attached to the `SummarizedExperiment`
objects. Only a dbGaP repository key is required to download, process, and
attach the protected metadata. See <https://dbgap.ncbi.nlm.nih.gov/aa> for
details on the protected metadata available, and how to get a key. 
The `attach_dbGap()` function takes a HMP16SData `SummarizedExperiment`
objects as its first argument and a path to your dbGaP repository key as its
second argument, and returns another `SummarizedExperiment` also containing the 
protected metadata. For example:

```{r, eval=FALSE}
V13_stool_protected <- V13_stool %>%
    attach_dbGaP("~/prj_12146.ngc")
```

The new `SummarizedExperiment` has protected metadata as additional columns in 
the colData. All open and protected metadata are returned by:

```{r, eval=FALSE}
colData(V13_stool_protected)
```

## Taxonomic Identifier Translation

`r BiocStyle::Biocpkg("HMP16SData")` provides methods to translate 
taxonomic identifiers to NCBI taxonomy standard. Like `attach_dbGaP`, the `translate_identifiers` 
function is endomorphic, taking and returning a `SummarizedExperiment` object. A
column, `ncbi_taxonomy`, is added to `rowData` with the following syntax:
n__NCBI Taxonomy ID, pk__superkingdom, k__kingdom, bk__subkingdom, p__phylum,
bp__subphylum, c__class, o__order, f__family, g__genus, s__species. Taxonomic
levels are pipe delimited and represented as a character string. Translation of
a single identifier is shown below for the purpose of brevity, but many
identifiers could be translated. Just be aware that translation of taxonomic
identifiers uses the Global Names Recognition and Discovery API and the
translation of many identifiers will take a long time.

The following example translates the identifiers of the 4th row only, then returns
the new `rowData`:

```{r, eval=FALSE}
V13_stool[4, ] %>%
    translate_identifiers() %>%
    rowData()
```

## External Data Comparison

Data from the HMP study is likely useful as a comparison or control group, a 
_"standard candle"_ of sorts, in the context of other human microbiome studies.
As such, `r BiocStyle::Biocpkg("HMP16SData")` provides a method, `as_phyloseq`,
to coerce `SummarizedExperiment` class objects to `phyloseq` class objects. The
`r BiocStyle::Biocpkg("phyloseq")` package provides an extensive suite of 
methods to analyze microbiome data and many potential comparisons could be made.

```{r, eval=FALSE}
V13_stool_phyloseq <-
    V13_stool %>%
    as_phyloseq()
```

When plotting `phyloseq` objects it is necessary to remove taxa that sum to zero
(i.e. have no relative abundance) to avoid warning messages. The 
`r BiocStyle::CRANpkg("magrittr")` package provides the most eloquent syntax to
do so – allowing piping between functions. See the 
`r BiocStyle::CRANpkg("magrittr")` vignette for more information.

```{r, eval=FALSE}
V13_stool_phyloseq %<>%
    taxa_sums() %>% 
    is_greater_than(0) %>%
    prune_taxa(V13_stool_phyloseq)
```

With zero-abundance taxa now pruned, exploratory plots can easily be made. For 
example, a plot of three alpha diversity measures can be made in three lines.
This concept could be extended to compare and external dataset.

```{r, eval=FALSE}
richness_measures <-
    c("Observed", "Shannon", "Simpson")
    
V13_stool_phyloseq %>%
    plot_richness(x = "sex", color = "sex", measures = richness_measures) +
    theme(legend.position = "none")
```

Again, `r BiocStyle::Biocpkg("phyloseq")` provides excellent plots with little
effort and the concept could be extended to compare an external dataset.

```{r, eval=FALSE}
V13_stool_phyloseq %>%
    plot_net(color = "sex")
```

## Frequency Table Generation

Sometimes it is desireable to provide a quick summary of key demographic 
variables and to make the process easier `r BiocStyle::Biocpkg("HMP16SData")` 
provides a function `table_one`. It returns a `data.frame` that has been 
transformed to make a publication ready table and when used in conjunction with
the `r BiocStyle::Githubpkg("schifferl/kableOne")` package can produce an `HTML`
table for vignettes. Feel free to use the `kable_one` function elsewhere but be
aware that the `r BiocStyle::Githubpkg("schifferl/kableOne")` package is _very_
beta.

```{r, message=FALSE}
V13 <-
    V13() %>%
    table_one()

V35 <-
    V35() %>%
    table_one()

kable_one(V13, V35)
```

# Limitations

Data from variable regions 1–3 and 3–5 are not comparable to everything, be 
careful with assumptions and conduct analysis that is methodologically sound.

# Session Information

```{r}
sessionInfo()
```

