---
title: "HMP16SData"
date: '`r format(Sys.Date(), "%B %e, %Y")`'
author:
- name: Lucas Schiffer
  affiliation: >
    Graduate School of Public Health and Health Policy, City University of New
    York, New York, NY
- name: Rimsha Azhar
  affiliation: >
    Graduate School of Public Health and Health Policy, City University of New
    York, New York, NY
- name: Marcel Ramos
  affiliation: >
    Graduate School of Public Health and Health Policy, City University of New
    York, New York, NY
- name: Ludwig Geistlinger
  affiliation: >
    Graduate School of Public Health and Health Policy, City University of New
    York, New York, NY
- name: Levi Waldron
  affiliation: >
    Graduate School of Public Health and Health Policy, City University of New
    York, New York, NY
package: HMP16SData
abstract: >
    HMP16SData is a Bioconductor ExperimentData package of the Human Microbiome
    Project (HMP) 16S rRNA sequencing data for variable regions 1–3 and 3–5. Raw
    data files are provided in the package as downloaded from the HMP Data
    Analysis and Coordination Center website. Processed data is provided as
    SummarizedExperiment objects via ExperimentHub.
output:
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{HMP16SData}
    %\VignetteEncoding{UTF-8}
    %\VignetteEngine{knitr::rmarkdown}
editor_options: 
    chunk_output_type: console
---

# Prerequisites

The following `r BiocStyle::CRANpkg("knitr")` options will be used in this 
vignette to provide the most useful and concise output.

```{r}
knitr::opts_chunk$set(message = FALSE)
```

The following packages will be used in this vignette to provide demonstrative
examples of what a user might do with `r BiocStyle::Biocpkg("HMP16SData")`.

```{r}
library(HMP16SData)
library(phyloseq)
library(magrittr)
library(dplyr)
library(ggplot2)
```

Pipe operators from the `r BiocStyle::CRANpkg("magrittr")` package are used in 
this vignette to provide the most eloquent and concise syntax. See the
`r BiocStyle::CRANpkg("magrittr")` vignette if the syntax is unclear.

# Introduction

`r BiocStyle::Biocpkg("HMP16SData")` is a Bioconductor ExperimentData package of
the Human Microbiome Project (HMP) 16S rRNA sequencing data for variable regions
1–3 and 3–5. Raw data files are provided in the package as downloaded from the
[HMP Data Analysis and Coordination Center](https://tinyurl.com/y7ev836z).
Processed data is provided as `r BiocStyle::Biocpkg("SummarizedExperiment")`
objects via `r BiocStyle::Biocpkg("ExperimentHub")`.
    
`r BiocStyle::Biocpkg("HMP16SData")` can be installed using 
`r BiocStyle::Biocpkg("BiocInstaller")` as follows:

```{r, eval=FALSE}
BiocInstaller::biocLite("HMP16SData")
```

Once installed, `r BiocStyle::Biocpkg("HMP16SData")` provides two functions to
access data – one for variable region 1–3 and another for variable region 3–5. 
When called, as follows, the functions will download data from an 
`r BiocStyle::Biocpkg("ExperimentHub")` Amazon S3 (Simple Storage Service) 
bucket over `https` or load data from a local cache.

```{r}
V13()
V35()
```

The two datasets are represented as `SummarizedExperiment` objects, a standard 
Bioconductor class that is amenable to subsetting and analysis. To maintain 
breveity, details of the `SummarizedExperiment` class are not outlined here but 
the `r BiocStyle::Biocpkg("SummarizedExperiment")` package provides an excellent
vignette.

# Features

## Frequency Table Generation

Sometimes it is desireable to provide a quick summary of key demographic 
variables and to make the process easier `r BiocStyle::Biocpkg("HMP16SData")` 
provides a function `table_one`. It returns a `data.frame` that has been 
transformed to make a publication ready table and when used in conjunction with
the `r BiocStyle::Githubpkg("schifferl/kableOne")` package can produce an `HTML`
table for vignettes. Feel free to use the `kable_one` function elsewhere but be
aware that the `r BiocStyle::Githubpkg("schifferl/kableOne")` package is _very_
beta.
<!-- remove eval=FALSE when methods are implemented -->
```{r, eval=FALSE}
V13() %>%
    table_one(variables = "SEX")
```

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque ipsum
ante, congue vitae ultricies at, congue id libero. Nulla blandit elit in tortor
malesuada, a egestas tortor finibus. Nulla aliquam arcu feugiat ornare eleifend.

<!-- remove eval=FALSE when methods are implemented -->
```{r, eval=FALSE}
V13 <- V13()
V35 <- V35()

list(V13, V35) %>%
    table_one() %>%
    kable_one()
```

## Straightforward Subsetting

The `SummarizedExperiment` container provides for straightforward subsetting by
data or metadata variables using either the `subset` function or `[` methods –
see the `r BiocStyle::Biocpkg("SummarizedExperiment")` vignette for additional
details. Shown below, the variable region 1–3 dataset is subset to include only
stool samples.

```{r}
V13_stool <-
    V13() %>%
    subset(select = HMP_BODY_SUBSITE == "Stool")

V13_stool
```

## Protected Metadata Attachment

The HMP study has extensive protected metadata available through the National
Center for Biotechnology Information's (NCBI) database of Genotypes and
Phenotypes (dbGaP) that can readily be attached to the `SummarizedExperiment`
objects. Only a dbGaP repository key is required to download, process, and
attach the protected metadata. See <https://dbgap.ncbi.nlm.nih.gov/aa> for
details on the protected metadata available, and how to get a key.
The `attach_dbGap()` function takes a `r BiocStyle::Biocpkg("HMP16SData")`
`SummarizedExperiment` objects as its first argument and a path to your dbGaP
repository key as its second argument, and returns another
`SummarizedExperiment` also containing the protected metadata. For example:

```{r, eval=FALSE}
V13_stool_protected <- 
    V13_stool %>%
    attach_dbGaP("~/prj_12146.ngc")
```

The new `SummarizedExperiment` has protected metadata as additional columns in 
the colData. All open and protected metadata are returned by:

```{r, eval=FALSE}
colData(V13_stool_protected)
```

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque ipsum
ante, congue vitae ultricies at, congue id libero. Nulla blandit elit in tortor
malesuada, a egestas tortor finibus. Nulla aliquam arcu feugiat ornare eleifend.

```{r}
head(dictionary)
```

## Phyloseq Class Coercion

Data from the HMP study is likely useful as a comparison or control group, a 
_"standard candle"_ of sorts, in the context of other human microbiome studies.
As such, `r BiocStyle::Biocpkg("HMP16SData")` provides a method, `as_phyloseq`,
to coerce `SummarizedExperiment` class objects to `phyloseq` class objects. The
`r BiocStyle::Biocpkg("phyloseq")` package provides an extensive suite of 
methods to analyze microbiome data and many potential comparisons could be made.

```{r}
V13_stool_phyloseq <-
    V13_stool %>%
    as_phyloseq()
```

When plotting `phyloseq` objects it is necessary to remove taxa that sum to zero
(i.e. have no relative abundance) to avoid warning messages. The 
`r BiocStyle::CRANpkg("magrittr")` package provides the most eloquent syntax to
do so – allowing piping between functions. See the 
`r BiocStyle::CRANpkg("magrittr")` vignette for more information.

```{r}
V13_stool_phyloseq %<>%
    taxa_sums() %>% 
    is_greater_than(0) %>%
    prune_taxa(V13_stool_phyloseq)
```

With zero-abundance taxa now pruned, exploratory plots can easily be made. For 
example, a plot of three alpha diversity measures can be made in three lines.

```{r}
richness_measures <- c("Observed", "Shannon", "Simpson")
    
V13_stool_phyloseq %>%
    plot_richness(x = "SEX", color = "SEX", measures = richness_measures) +
    theme(legend.position = "none")
```

# External Data Comparison

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque ipsum
ante, congue vitae ultricies at, congue id libero. Nulla blandit elit in tortor
malesuada, a egestas tortor finibus. Nulla aliquam arcu feugiat ornare eleifend.

```{r}
melted <-
    V13_stool_phyloseq %>%
    tax_glom(taxrank = "PHYLUM") %>%
    psmelt()
```

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque ipsum
ante, congue vitae ultricies at, congue id libero. Nulla blandit elit in tortor
malesuada, a egestas tortor finibus. Nulla aliquam arcu feugiat ornare eleifend.

```{r}
relative_abundance <-
    melted %>%
    rename(Phylum = PHYLUM) %>%
    group_by(Sample) %>%
    mutate(`Relative Abundance` = Abundance / sum(Abundance)) %>%
    arrange(Phylum, -`Relative Abundance`) %>%
    group_by(Phylum) %>%
    mutate(phylum_rank = sum(`Relative Abundance`)) %>%
    arrange(-phylum_rank) %>%
    select(Sample, Phylum, `Relative Abundance`) %>%
    ungroup()
```

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque ipsum
ante, congue vitae ultricies at, congue id libero. Nulla blandit elit in tortor
malesuada, a egestas tortor finibus. Nulla aliquam arcu feugiat ornare eleifend.

```{r}
top_eight_phyla <-
    relative_abundance %$%
    unique(Phylum) %>%
    as.character() %>%
    extract(1:8)
```

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque ipsum
ante, congue vitae ultricies at, congue id libero. Nulla blandit elit in tortor
malesuada, a egestas tortor finibus. Nulla aliquam arcu feugiat ornare eleifend.

```{r}
relative_abundance %<>%
    filter(Phylum %in% top_eight_phyla)
```

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque ipsum
ante, congue vitae ultricies at, congue id libero. Nulla blandit elit in tortor
malesuada, a egestas tortor finibus. Nulla aliquam arcu feugiat ornare eleifend.

```{r}
sample_order <-
    relative_abundance %$%
    unique(Sample)
```

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque ipsum
ante, congue vitae ultricies at, congue id libero. Nulla blandit elit in tortor
malesuada, a egestas tortor finibus. Nulla aliquam arcu feugiat ornare eleifend.

```{r}
phylum_order <-
    relative_abundance %$%
    unique(Phylum)
```

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque ipsum
ante, congue vitae ultricies at, congue id libero. Nulla blandit elit in tortor
malesuada, a egestas tortor finibus. Nulla aliquam arcu feugiat ornare eleifend.

```{r}
bang_wong_colors <-
    c(
        "#CC79A7",
        "#D55E00",
        "#0072B2",
        "#F0E442",
        "#009E73",
        "#56B4E9",
        "#E69F00",
        "#000000"
    )
```

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque ipsum
ante, congue vitae ultricies at, congue id libero. Nulla blandit elit in tortor
malesuada, a egestas tortor finibus. Nulla aliquam arcu feugiat ornare eleifend.

```{r}
relative_abundance %>%
    mutate(Sample = factor(Sample, sample_order)) %>%
    mutate(Phylum = factor(Phylum, phylum_order)) %>%
    ggplot(aes(Sample, `Relative Abundance`, fill = Phylum)) +
    geom_bar(stat = "identity", position = "fill", width = 1) +
    scale_fill_manual(values = bang_wong_colors) +
    theme_minimal() +
    theme(axis.text.x = element_blank(), axis.title.x = element_blank(),
          axis.title.y = element_blank(), panel.grid = element_blank(),
          legend.position = "bottom", legend.title = element_blank()) +
    ggtitle("Phylum-Level Relative Abundance", "HMP16SData Stool Samples")
```

# Limitations

Data from variable regions 1–3 and 3–5 are not comparable to everything, be 
careful with assumptions and conduct analysis that is methodologically sound.

# Session Information

```{r}
sessionInfo()
```

